Infinite depth configuration

GravlLift opened the pull request:
This PR allows for more targeted configuration of state, removing the restriction that only top-level data can be stored. For instance, this is now a valid configuration:

```
export function localStorageSyncReducer(
  reducer: ActionReducer<any>
): ActionReducer<any> {
  return localStorageSync({
    keys: [
      {
        account: [
          { settings: [{ settingsGroup1: ['oneSettingInParticular'] }] },
        ],
      },
    ]
  })(reducer);
}
```
The above configuration would result in a localstorage value under the key `account` that looked like so, ignoring all other values in account, settings, or settingsGroup1:
```
{
  "settings": {
    "settingsGroup1": {
      "oneSettingInParticular": "value of setting"
    }
  }
}
```

Upgraded TS to 3.9.7
Added .editorconfig

BBlackwo commented on spec/index_spec.ts:
```diff
@@ -492,39 +495,40 @@ describe('ngrxLocalStorage', () => {
+        });
+
+        const action = { type: INIT_ACTION };
+
```
Please add a test for multiple levels deep config

BBlackwo commented:
Hi Jason @GravlLift,

Thanks so much for the PR! This is great. I love how you've cleaned up the code and added types.

Can you please:
- Fix up the comments above (type consistency and adding a test)
- Update the docs with an example config multiple levels deep.
- Rebase off the lastest master and resolve the conflicts
- Update your commit messages to follow the new required [commit message format ](https://github.com/btroncone/ngrx-store-localstorage/blob/master/CONTRIBUTING.md#commit). You can do this with an interactive rebase.

GravlLift commented on src/index.ts:
```diff
@@ -5,309 +5,311 @@ const UPDATE_ACTION = '@ngrx/store/update-reducers';
+    serialize?: (state: any) => any;
+    reviver?: (key: string, value: any) => any;
+    replacer?: ((key: string, value: any) => any) | string[];
+    encrypt?(message: string): string;
```
I've fixed this up. On a related note, [this sort of styling rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/method-signature-style.md) is available as part of [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint), the successor to the now deprecated tslint that this project is using. You might consider upgrading.

BBlackwo replied:
Good to know! Thanks for the heads up. I've added to this issue to upgrade to TSLint #169 as well as some other build changes.

GravlLift commented:
All set @BBlackwo.

- [x] Fix up the comments above (type consistency and adding a test)
- [x] Update the docs with an example config multiple levels deep.
- [x] Rebase off the lastest master and resolve the conflicts
- [x] Update your commit messages to follow the new required commit message format .

BBlackwo commented on src/lib/index.ts:
```diff
@@ -5,309 +5,311 @@ const UPDATE_ACTION = '@ngrx/store/update-reducers';
 }
+
+interface Options {
+    serialize?: (state: any) => any;
```
This is missing the `deserialize` function.

GravlLift replied:
Added.

BBlackwo reviewed (changes requested):
Thanks for addressing all of those things! I've just got a few more minor comments if you can have a look please.

BBlackwo commented on README.md:
````diff
@@ -96,7 +96,24 @@ An interface defining the configuration attributes to bootstrap `localStorageSyn
-Usage: `localStorageSync({keys: ['todos', 'visibilityFilter'], storageKeySerializer: (key) => 'cool_' + key, ... })`. In this example `Storage` will use keys `cool_todos` and `cool_visibilityFilter` keys to store `todos` and `visibilityFilter` slices of state). The key itself is used by default - `(key) => key`.
+### Usage
+#### Key Prefix
+```
````
Thanks for cleaning this up! Can you add the syntax here and in the below code block please.
```suggestion
```ts
```

GravlLift replied:
That's pretty cool, didn't know about that markdown feature!

BBlackwo replied:
Yeah it's really nice. I suppose it might not work in every markdown editor, but most support it :)

BBlackwo commented on README.md:
````diff
@@ -96,7 +96,24 @@ An interface defining the configuration attributes to bootstrap `localStorageSyn
+    ],
+});
+```
+In this example, feature1.slice11.slice11_1, feature1.slice14.slice14_2, and feature2.slice21 will be synced to localStorage.feature1 and localStorage.feature2.
````
Might be easier to read as code
```suggestion
In this example, `feature1.slice11.slice11_1`, `feature1.slice14.slice14_2`, and `feature2.slice21` will be synced to `localStorage.feature1` and `localStorage.feature2`.
```

GravlLift replied:
Added

GravlLift commented on spec/index_spec.ts:
```diff
@@ -492,39 +499,72 @@ describe('ngrxLocalStorage', () => {
+    });
+
+    it('should save targeted infinite depth to localStorage', () => {
+        // Configure to omly save feature1.slice11.slice11_1 and feature2.slice12,
```
Fixed.

GravlLift commented on src/lib/index.ts:
```diff
@@ -5,309 +5,311 @@ const UPDATE_ACTION = '@ngrx/store/update-reducers';
+function createStateSlice(existingSlice: any, filter: (string | number | KeyConfiguration | Options)[]) {
+    return filter.reduce((memo: { [x: string]: any; [x: number]: any }, attr: KeyConfiguration | Options) => {
+        if (typeof attr === 'string' || typeof attr === 'number') {
+            memo[attr] = existingSlice == null ? null : existingSlice[attr];
```
As I was thinking about this a bit more, it occurred to me that if a property is undefined in the state, we would not want to define it in local storage either (as opposed to setting it null). I modified the code to reflect that.

That said, the null terminator suggestion seems to be a good idea.

BBlackwo replied:
That makes sense. Thanks.

GravlLift commented on src/lib/index.ts:
```diff
@@ -5,309 +5,311 @@ const UPDATE_ACTION = '@ngrx/store/update-reducers';
-
-        if (stateSlice === 'null' || isObjectRegex.test(stateSlice.charAt(0))) {
-          raw = JSON.parse(stateSlice, reviver);
+function createStateSlice(existingSlice: any, filter: (string | number | KeyConfiguration | Options)[]) {
```
Added.

GravlLift commented:
Updated according to comments.

On an unrelated note, I'm not sure exactly how the commit/release note tool works, but it might be useful to apply that commit formatting restriction only to the master branch, and then simply squash PRs into a single commit with the correct formatting. That way you can have messy PR branches that you can update and commit as much as you like without restriction, but that mess gets condensed and removed into a single commit with a useful message when it goes to master. I'm rewriting the same commit over and over in this branch with force push anyways, so in essence you're losing that branch history regardless.

Just a suggestion.

BBlackwo commented:
Thanks for fixing those up 👌

>it might be useful to apply that commit formatting restriction only to the master branch, and then simply squash PRs into a single commit with the correct formatting.

That's a good idea. It looks like semantic release [might not work well with squashed commits](https://semantic-release.gitbook.io/semantic-release/support/troubleshooting#squashed-commits-are-ignored-by-semantic-release), sadly. But I'll keep it in mind!

BBlackwo reviewed (approved):
Thanks so much for this enhancement, Jason @GravlLift!

BBlackwo commented:
:tada: This PR is included in version 10.1.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/ngrx-store-localstorage/v/10.1.0)
- [GitHub release](https://github.com/btroncone/ngrx-store-localstorage/releases/tag/v10.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

BBlackwo commented:
Sorry it took so long for me to merge this in, btw. Appreciate your patience and continuing to address my comments months later.
